numpy/_core/src/umath/ufunc_object.c

```diff
@@ -935,6 +935,19 @@ convert_ufunc_arguments(PyUFuncObject *ufunc,
         out_op_DTypes[i] = NPY_DTYPE(PyArray_DESCR(out_op[i]));
         Py_INCREF(out_op_DTypes[i]);
 
+        if (nin == 1) {
+            /*
+             * TODO: If nin == 1 we don't promote!  This has exactly the effect
+             *       that right now integers can still go to object/uint64 and
+             *       their behavior is thus unchanged for unary ufuncs (like
+             *       isnan).  This is not ideal, but pragmatic...
+             *       We should eventually have special loops for isnan and once
+             *       we do, we may just deprecate all remaining ones (e.g.
+             *       `negative(2**100)` not working as it is an object.)
```

Comment:
It would be nice if this comment mentioned it was a TODO from the NEP 50 work, for context

Reply:
I have it in the tracking issue...

Reply:
I'm thinking of a reader who doesn't have context about NEP 50 and wouldn't be aware of a tracking issue.

Reply:
Ah, sorry, somehow was thinking about creating an issue.